homeassistant/components/harmony/data.py

```diff
@@ -0,0 +1,333 @@
+        )
+
+    @property
+    def activity_names(self):
+        """Names of all the remotes activities."""
+        activity_infos = self._client.config.get("activity", [])
+        activities = [activity["label"] for activity in activity_infos]
+
+        # Remove both ways of representing PowerOff
+        if None in activities:
+            activities.remove(None)
+        if ACTIVITY_POWER_OFF in activities:
+            activities.remove(ACTIVITY_POWER_OFF)
+
+        return activities
+
+    @property
+    def device_names(self):
+        """Names of all of the devices connected to the hub."""
+        device_infos = self._client.config.get("device", [])
+        devices = [device["label"] for device in device_infos]
+
+        return devices
+
+    @property
+    def name(self):
+        """Return the Harmony device's name."""
+        return self._name
+
+    @property
+    def unique_id(self):
+        """Return the Harmony device's unique_id."""
+        return self._unique_id
+
+    @property
+    def available(self) -> bool:
+        """Return if connected to the hub."""
+        return self._available
+
+    @property
+    def current_activity(self) -> tuple:
+        """Return the current activity tuple."""
+        return self._client.current_activity
+
+    @property
+    def json_config(self) -> dict:
+        """Return the hub config as json."""
+        if self._client.config is None:
+            return None
+        return self._client.json_config
+
+    def device_info(self, domain: str):
+        """Return hub device info."""
+        model = "Harmony Hub"
+        if "ethernetStatus" in self._client.hub_config.info:
+            model = "Harmony Hub Pro 2400"
+        return {
+            "identifiers": {(domain, self.unique_id)},
+            "manufacturer": "Logitech",
+            "sw_version": self._client.hub_config.info.get(
+                "hubSwVersion", self._client.fw_version
+            ),
+            "name": self.name,
+            "model": model,
+        }
+
+    async def connect(self) -> bool:
+        """Connect to the Harmony Hub."""
+        _LOGGER.debug("%s: Connecting", self._name)
+        try:
+            if not await self._client.connect():
+                _LOGGER.warning("%s: Unable to connect to HUB", self._name)
+                await self._client.close()
+                return False
+        except aioexc.TimeOut:
+            _LOGGER.warning("%s: Connection timed-out", self._name)
+            return False
+        return True
+
+    async def shutdown(self):
+        """Close connection on shutdown."""
+        _LOGGER.debug("%s: Closing Harmony Hub", self._name)
+        try:
+            await self._client.close()
+        except aioexc.TimeOut:
+            _LOGGER.warning("%s: Disconnect timed-out", self._name)
+
+    async def async_start_activity(self, activity: str):
+        """Start an activity from the Harmony device."""
+
+        if activity:
```

Comment:
Many of these long methods for starting activities, sending commands, etc were directly moved from the remote code with very minor modifications to support their new location.